### PR TITLE
many: add spread testing using GCE backend for ubuntu-core-initramfs

### PR DESCRIPTION
**Pull request by anonymouse64**
This sets up the project to be testable via spread, both locally via nested QEMU
VM's, as well as in the cloud with GCE nested VM's.

There are some TODO's still around simplifying this even more and getting rid of
more cruft from snapd's setup, but this is sufficient right now to get a nested
UC20 VM with our version of the initramfs in it for testing.

To run the tests locally via qemu, first install qemu, then get spread via 

```
$ sudo apt update && sudo apt install -y qemu-kvm autopkgtest
$ curl https://niemeyer.s3.amazonaws.com/spread-amd64.tar.gz | tar -xz -C $GOPATH/bin
```

Then make your qemu spread images dir with

```
$ mkdir -p ~/.spread/qemu
$ cd ~/.spread/qemu
```

and make a suitable spread image with:

```
$ autopkgtest-buildvm-ubuntu-cloud -r focal
$ mv autopkgtest-focal-amd64.img ubuntu-20.04-64.img
```

then finally you can run spread tests like so:

```
$ spread qemu-nested
2020-08-19 13:31:03 Project content is packed for delivery (9.00MB).
2020-08-19 13:31:03 Sequence of jobs produced with -seed=1597861863
2020-08-19 13:31:03 If killed, discard servers with: spread -reuse-pid=86651 -discard
2020-08-19 13:31:03 Allocating qemu-nested:ubuntu-20.04-64...
2020-08-19 13:31:03 Serial and monitor for qemu-nested:ubuntu-20.04-64 available at ports 59414 and 59514.
2020-08-19 13:31:03 Waiting for qemu-nested:ubuntu-20.04-64 to make SSH available...
2020-08-19 13:31:04 Allocated qemu-nested:ubuntu-20.04-64.
2020-08-19 13:31:04 Connecting to qemu-nested:ubuntu-20.04-64...
2020-08-19 13:31:22 Connected to qemu-nested:ubuntu-20.04-64 at localhost:59314.
2020-08-19 13:31:22 Sending project content to qemu-nested:ubuntu-20.04-64...
2020-08-19 13:31:23 Preparing qemu-nested:ubuntu-20.04-64 (qemu-nested:ubuntu-20.04-64)...
2020-08-19 13:45:03 Executing qemu-nested:ubuntu-20.04-64:tests/spread/basic:noenc (qemu-nested:ubuntu-20.04-64) (1/4)...
2020-08-19 13:53:49 Restoring qemu-nested:ubuntu-20.04-64:tests/spread/basic:noenc (qemu-nested:ubuntu-20.04-64)...
2020-08-19 13:53:49 Executing qemu-nested:ubuntu-20.04-64:tests/spread/basic:fullenc (qemu-nested:ubuntu-20.04-64) (2/4)...
2020-08-19 14:04:44 Restoring qemu-nested:ubuntu-20.04-64:tests/spread/basic:fullenc (qemu-nested:ubuntu-20.04-64)...
2020-08-19 14:04:45 Executing qemu-nested:ubuntu-20.04-64:tests/spread/basic:onlysecboot (qemu-nested:ubuntu-20.04-64) (3/4)...
2020-08-19 14:13:36 Restoring qemu-nested:ubuntu-20.04-64:tests/spread/basic:onlysecboot (qemu-nested:ubuntu-20.04-64)...
2020-08-19 14:13:36 Executing qemu-nested:ubuntu-20.04-64:tests/spread/basic:onlytpm (qemu-nested:ubuntu-20.04-64) (4/4)...
2020-08-19 14:24:44 Restoring qemu-nested:ubuntu-20.04-64:tests/spread/basic:onlytpm (qemu-nested:ubuntu-20.04-64)...
2020-08-19 14:24:45 Discarding qemu-nested:ubuntu-20.04-64...
2020-08-19 14:24:45 Successful tasks: 4
2020-08-19 14:24:45 Aborted tasks: 0
```

This takes approximately an hour for me on my well spec'd desktop machine, largely because we can't use nested KVM accelerated VM's, so the L2 nested VM runs significantly slower.

To run the tests via GCE, you would just run

```
$ spread google-nested
```

but to do so you need GCE credentials, see https://github.com/snapcore/spread#google

### CODE REVIEW

File: tests/lib/prepare-uc20.sh [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+# extract the kernel snap, including extracting the initrd from the kernel.efi
+kerneldir=/tmp/kernel-workdir
+"$TESTSLIB/repack-kernel.sh" extract upstream-pc-kernel.snap $kerneldir
+
+# extract snap-bootstrap from the snapd snap and put into the kernel work dir
+cp /tmp/snap-bootstrap-from-snapd-snap $kerneldir/skeleton/main/usr/lib/snapd/snap-bootstrap
+
+# copy the skeleton from our installed ubuntu-core-initramfs into the initrd 
+# skeleton for the kernel snap
+cp -ar /usr/lib/ubuntu-core-initramfs/main/* "$kerneldir/skeleton/main"
```
Comment (anonymouse64) [inline tests/lib/prepare-uc20.sh:189]: this in effect means that we are using the same version of kernel modules from the snap that were already there, the same kernel, a new snap-bootstrap from edge channel of snapd, and a new version of the ubuntu-core-initramfs layout from spread/the PR under test.

I seem to remember there being some thinking that what we were doing in snapd was not great and that we should be using some other location/source for the kernel modules or something, but perhaps I am just misremembering

File: tests/lib/repack-kernel.sh [hunk 2]
```diff
@@ -0,0 +1,233 @@
+#!/bin/bash
```
Comment (anonymouse64) [inline tests/lib/repack-kernel.sh:1]: ideally $SOMEDAY this whole script becomes a command on the ubuntu-core-initramfs snap

File: tests/spread/basic/task.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  # Now formally wait for seeding
+  execute_remote sudo snap wait system seed.loaded
+
+  # We should be in run mode
+  execute_remote cat /proc/cmdline | MATCH snapd_recovery_mode=run
+
+  # We should have a modeenv
+  execute_remote test -f /var/lib/snapd/modeenv
+
+  # TODO: what else qualifies a "successful boot"?
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply tests/spread/basic/task.yaml:33]: I should mention that we have tests for recover mode via spread in snapd, but the setup is complicated, but we do not have any tests for install mode, so setting up tests for install mode is very hard and setting up tests for recover mode is just "kinda hard".

File: tests/lib/prepare-uc20.sh [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+if getent group systemd-journal >/dev/null; then
+    usermod -G systemd-journal -a test
+    id test | MATCH systemd-journal
+fi
+
+echo 'test ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+
+# get the model
+# TODO: should this be a variant to test multiple grades i.e. signed and secured?
```
Comment (anonymouse64) [inline tests/lib/prepare-uc20.sh:44]: looking at this again, it occurs to me that we can't specify unasserted snaps for signed or secured models, which means we couldn't inject users into the VM's for those grades, so probably this TODO should just go away and we are stuck with dangerous only grades.

restricting ourselves to just dangerous also means we could just use cloud-init instead of the whole special user setup too :thinking:

**Comment by xnox** [comment]
> This takes approximately an hour for me on my well spec'd desktop machine, largely because we can't use nested KVM accelerated VM's, so the L2 nested VM runs significantly slower.

Why are nested VMs used locally? Can spread not launch the VM under test from my local machine directly? Can environment be fixed to allow that?

File: spread.yaml [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  - "*.img"
+
+prepare: |
+  # NOTE: This part of the code needs to be in spread.yaml as it runs before
+  # the rest of the source code (including the tests/lib directory) is
+  # around. The purpose of this code is to fix some connectivity issues setup 
+  # spread functions to be more usable outside of the explicit task.yaml 
+  # sections.
+
+  # apt update is hanging on security.ubuntu.com with IPv6, prefer IPv4 over IPv6
```
Comment (xnox) [inline spread.yaml:52]: huh, no way. Is there a bug about this? Are you sure that nested VMs have correct ipv6 passthrough setup?

File: spread.yaml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      ENABLE_SECURE_BOOT/fullenc: true
+      # only secboot does not have a TPM
+      ENABLE_TPM/onlysecboot: false
+      ENABLE_SECURE_BOOT/onlysecboot: true
+      # only tpm does not have secure boot
+      ENABLE_TPM/onlytpm: true
+      ENABLE_SECURE_BOOT/onlytpm: false
+      # no enc does not have secure boot or tpm
+      ENABLE_TPM/noenc: false
+      ENABLE_SECURE_BOOT/noenc: false
```
[the remark this answers is not in the document]
Comment (xnox) [reply spread.yaml:121]: I don't like using the name "secboot" in the variant name, especially in the one that doesn't refer to what we call secboot (measured and secured boot with FDE).

Also 3 out of 4 variants are effectively the same. Because without either secureboot or without tpm or without both, we just do plain boot.

Can't we reduce this to just two variants, always with a tpm, and one variant with secureboot on, second variant with secureboot off?

Or is there something special about testting secureboot-off with tpm or secureboot-on without tpm? I don't think there is any different code paths for those, are there?

File: tests/lib/nested.sh [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    PARAM_SERIAL="-serial file:${WORK_DIR}/serial.log"
+    PARAM_TPM=""
+
+    ATTR_KVM=""
+    if [ "$ENABLE_KVM" = "true" ]; then
+        ATTR_KVM=",accel=kvm"
+        # CPU can be defined just when kvm is enabled
+        PARAM_CPU="-cpu host"
+        # Increase the number of cpus used once the issue related to kvm and ovmf is fixed
+        # https://bugs.launchpad.net/ubuntu/+source/kvm/+bug/1872803
```
Comment (xnox) [inline tests/lib/nested.sh:91]: Above was fixed in june, so can we try again with more CPU cores nested in Google cloud?

File: tests/lib/prepare-uc20.sh [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    cp ../*.deb "$SETUPDIR"
+)
+
+# install ubuntu-core-initramfs here so the repack-kernel.sh uses the version of
+# ubuntu-core-initramfs we built here
+apt install -yqq "$SETUPDIR"/ubuntu-core-initramfs*.deb
+
+# now download snap dependencies 
+# TODO:UC20: when should some of these things start tracking stable ?
+snap download pc-kernel --channel=20/edge --basename=upstream-pc-kernel
```
Comment (xnox) [inline tests/lib/prepare-uc20.sh:65]: ooooh, i didn't know about `--basename` arg.

File: tests/lib/prepare-uc20.sh (hunk 1 above)
Comment (xnox) [reply tests/lib/prepare-uc20.sh:190]: there was, and when we tried that, we found bugs =) which were not fixed yet.

It would be nicer to execute ubuntu-core-initramfs to create and sign kernel.efi in one go, for the kernel/modules extracted from the snap. But that's not possible yet.

Also not sure if during this test we want to use existing snap-bootstrap from inside the kernel-snap, the one from snapd edge, or the one that is "current" / stable / vendored inside ubuntu-core-initramfs.

File: tests/spread/basic/task.yaml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  # We should be in run mode
+  execute_remote cat /proc/cmdline | MATCH snapd_recovery_mode=run
+
+  # We should have a modeenv
+  execute_remote test -f /var/lib/snapd/modeenv
+
+  # TODO: what else qualifies a "successful boot"? 
+  # * no failed systemd units?
+  # * all expected partitions mounted?
+  # * systemd not in degraded mode?
```
Comment (xnox) [inline tests/spread/basic/task.yaml:36]: LUKS has been setup for ubuntu-data, when expected.

TPM measurements were completed.

snapd has sealed things.

**Review by xnox — APPROVED** [review]
There are a few todo's that we should revisit, but this is mostly ok and helps automating some of our testing.

Should I have already expected a check executed on this PR? or does it need to be merged first before adding github actions to it?

**Comment by anonymouse64** [comment]
> > This takes approximately an hour for me on my well spec'd desktop machine, largely because we can't use nested KVM accelerated VM's, so the L2 nested VM runs significantly slower.
> 
> Why are nested VMs used locally? Can spread not launch the VM under test from my local machine directly? Can environment be fixed to allow that?

That would require changes to spread, it can be done, but then you have to deal with the fact that someone could be running spread on this system from their xenial machine (or even Fedora!) and not have new QEMU, and not have all the other dependencies, etc. and makes the output less predictable and potentially fails with random problems due to these dependencies. With this setup, all you need is spread + some qemu and it will always work

File: tests/spread/basic/task.yaml (hunk 9 above)
Comment (anonymouse64) [reply tests/spread/basic/task.yaml:36]: Sure, I will leave it to you to write such tests in followups, I just left this as a clear entry-point for foundations to take over testing and write whatever tests are appropriate

File: spread.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply spread.yaml:121]: Sure if you want to eliminate 2 of the variants to just have secure boot w/ TPM (and thus also full disk encryption), as well as the plain boot without any of this that's fine. I included them because there is code in snap-bootstrap that operates slightly differently depending on if we have just secure boot or just a TPM, etc. but we already have tests for this different behavior in snapd, so that setup doesn't need to be replicated here.

File: tests/lib/prepare-uc20.sh (hunk 1 above)
Comment (anonymouse64) [reply tests/lib/prepare-uc20.sh:190]: yes I also am not sure what the right thing to do is, you probably saw in the PR, but I actually changed it slightly after opening the PR to not stage snap-bootstrap from the snapd snap into the kernel snap's initramfs, but actually use the version of snap-bootstrap that comes from snapd from the snappy-dev/image PPA so that this test tests exactly what ubuntu-core-initramfs would do if it was built today.

that being said it would still be kinda nice if we have testing somewhere here like a nightly or something that ran with ubuntu-core-initramfs from main here, but also staged snap-bootstrap from edge snapd snap into it too to make sure that continues to work and doesn't regress as edge changes

**Comment by anonymouse64** [comment]
> Should I have already expected a check executed on this PR? or does it need to be merged first before adding github actions to it?

No, I spoke with Gustavo, he preferred to merge the spread setup first then get the keys setup, I will work on doing that now with him since we have merged this however.

File: tests/lib/nested.sh (hunk 7 above)
Comment (anonymouse64) [reply tests/lib/nested.sh:91]: sure go for it, but every time we try to use faster / better VM's in GCE it doesn't work for us in weird ways, so YMMV

File: spread.yaml (hunk 5 above)
Comment (anonymouse64) [reply spread.yaml:52]: well this was originally added to snapd's spread.yaml as of https://github.com/snapcore/snapd/pull/3418, so perhaps things are better now, I will leave it to you to look into removing this, if it works for you to remove it, then let us know and we'd be happy to remove the same from our spread.yaml :-)
